code401challenges/src/main/java/linkedLists/LinkedList.java

```diff
@@ -4,36 +4,76 @@
     Node<T> head; // head of list
 
     public LinkedList() {
+
         this.head = null;
     }
 
     // Method to insert a new node
-    public void Insert(T value) {
-        Node newNode = new Node(value, this.head);
-        this.head = newNode;
+    public void insert(T value) {
+        this.head = new Node<>(value, this.head);
 
     }
 
     //Append to end
-    public void Append(T value) {
-        Node<T> newNode = new Node<>(value);
-        newNode.next = null;
+    //Was mixing up the head and the current node so kept getting null values
+    public void append(T value) {
+        Node<T> newNode = new Node<>(value, this.head);
+        Node<T> current = this.head;
         // base case
-        if (head == null) {
-            head = newNode;
-        } else {
-            Node<T> last = head;
-            while (last.next != null) {
-                last = last.next;
+        if (current == null) {
+            current = newNode;
+            this.head = current;
+            return;
+        }
+        newNode.setNext(null);
+        while (current.getNext() != null) {
+            current = current.getNext();
+        }
+        current.setNext(newNode);
+    }
+
+
+    public void insertBefore(T targetVal, T newValue) {
+        Node<T> current = head;
+
+        if (current.value.equals(targetVal)) {
+            this.insert(newValue);
+            return;
+        }
+        while (true) {

```

Comment:
This totally works but I get nervous when passing in `true` as the conditional for a `while` loop. It's too easy to accidentally loop infinitely. Instead of creating a separate `if` statement to choose when to stop the loop, you could have moved the conditional logic from the `if` to the `while`:

`while (!current.next.value.equals(targetVal))`

This required a `!` at the front of the conditional to work, but this will stop the loop when you need it to and you wouldn't need the `if / else` at all.